File: Leetcode Problems/121.cpp
#include<iostream>
#include<vector>
using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        int mp = 0;
        int bb = prices[0];

        for(int i=1;i<n;i++){
            if(prices[i] > bb){
                mp = max(mp, prices[i] - bb); 
            }
            bb = min(bb, prices[i]);
        }
        return mp;
    }
};

int main() {
    Solution sol;
    vector<int> prices1 = {7, 1, 5, 3, 6, 4};
    vector<int> prices2 = {7, 6, 4, 3, 1};
    cout << sol.maxProfit(prices1) << endl;
    cout << sol.maxProfit(prices2) << endl; 
    return 0;
}

File: Leetcode Problems/121.py
from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        l = 0                                             # 2 Pointer Approach
        r = 1
        maxP = 0

        while(r < len(prices)):
            if(prices[l] < prices[r]):
                p = prices[r] - prices[l]
                maxP = max(p, maxP)
            else:
                l = r
            r += 1
        
        return maxP

if __name__ == "__main__":
        prices = [7, 1, 5, 3, 6, 4]
        solution = Solution()
        print(solution.maxProfit(prices))